official/keras_application_models/model_callbacks.py

```diff
@@ -0,0 +1,159 @@
+`every_n_steps` options in tensorflow Estimator hooks.
+"""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import time
+
+# pylint: disable=g-bad-import-order
+import tensorflow as tf
+# pylint: enable=g-bad-import-order
+
+from official.utils.logs import logger
+
+CALLBACKS = ["examplespersecondcallback", "loggingmetriccallback"]
+
+_METRICS_TO_LOG = {
+    "loss": "train_loss",
+    "acc": "train_accuracy",
+    "val_loss": "loss",
+    "val_acc": "accuracy"
+}
+
+
+class ExamplesPerSecondCallback(tf.keras.callbacks.Callback):
+  """ExamplesPerSecond callback.
+
+  This callback records the average examples per second during training.
+  """
+
+  def __init__(self, batch_size=None, epoch_size=None, batch_based=None,
+               epoch_based=None, metric_logger=None):
+    if (batch_based is None) == (epoch_based is None):
+      raise ValueError("Exactly one of batch_based "
+                       "and epoch_based should be provided.")
+
+    if batch_based and (batch_size is None):
+      raise ValueError("batch_size should be provided for "
+                       "batch_based benchmark.")
+
+    if epoch_based and (epoch_size is None):
+      raise ValueError("epoch_size should be provided for "
+                       "epoch_based benchmark.")
+
+    self._batch_based = batch_based
+    self._epoch_based = epoch_based
+    self._batch_size = batch_size
+    self._epoch_size = epoch_size
+
+    self._logger = metric_logger or logger.BaseBenchmarkLogger()
+
+  def on_train_begin(self, logs=None):
+    self._global_step = 0
+    # For batch_based
+    if self._batch_based:
+      self._train_time_batch_based = 0
+    else:  # For epoch_based
+      self._epochs = 0
+      self._train_time_epoch_based = 0
+
+  def on_batch_begin(self, batch, logs=None):
+    if self._batch_based:
+      self._time_start_batch_based = time.time()
+
+  def on_batch_end(self, batch, logs=None):
+    self._global_step += 1
+    if self._batch_based:
+      self._train_time_batch_based += time.time() - self._time_start_batch_based
+      examples_per_sec_batch_based = self._batch_size * (
+          self._global_step / self._train_time_batch_based)
+      self._logger.log_metric(
+          "examples_per_sec_batch_based",
+          examples_per_sec_batch_based,
+          global_step=self._global_step)
+
+  def on_epoch_begin(self, epoch, logs=None):
+    if self._epoch_based:
+      self._time_start_epoch_based = time.time()
+
+  def on_epoch_end(self, epoch, logs=None):
+    if self._epoch_based:
+      self._epochs += 1
+      self._train_time_epoch_based += time.time() - self._time_start_epoch_based
+      examples_per_sec_epoch_based = self._epoch_size * (
+          self._epochs / self._train_time_epoch_based)
+      self._logger.log_metric(
+          "examples_per_sec_epoch_based",
+          examples_per_sec_epoch_based,
+          global_step=self._global_step)
+
```

Comment:
I feel like only one of the epoch/batch choices should be necessary, especially for ex per seconds. Let's just do batch, and then not have epoch_based, and that simplifies the code a lot, and matches the use-case just fine.

Reply:
Yes, it will definitely simply the code here. As for our earlier discussions with Scott , the epoch/batch choices can provide different granularity of benchmark metrics. If the batch size is small and dataset is large, batch_based option will provide lots of values which may be not desired for benchmark. @qlzh727 any idea here?